0022-circuit-block.md

```diff
@@ -0,0 +1,153 @@
+        qargs = tuple(
+            {q: None for qarg in qargs for q in self._qbit_argument_conversion(qarg)}
+        )
+        cargs = tuple(
+            {b: None for carg in cargs for b in self._clbit_argument_conversion(carg)}
+        )
+
+        return self.append(Block(circuit, ...), qargs, cargs)
+```
+
+The logic in the `block` function above returns an `InstructionSet` containing a single `CircuitInstruction`, with the block as `operation` and the correct register's slices as `qubits` and `clbits`:
+```python
+block = QuantumCircuit(2)
+block.cx(1, 0)
+
+qreg1 = QuantumRegister(3, "qreg1")
+qreg2 = QuantumRegister(3, "qreg2")
+
+circuit = QuantumCircuit(qreg1, qreg2)
+circuit.s(qreg1[1])
+circuit.h(qreg2[1])
+my_block = circuit.block(block, [1, 5])
+# equivalent to: my_block = circuit.block(block, [qreg1[1], qreg2[2]])
+
+print(my_block[0])
+>>> "CircuitInstruction(operation=Instruction(name='block', num_qubits=2, num_clbits=0, params=[]), qubits=(Qubit(QuantumRegister(3, 'qreg1'), 1), Qubit(QuantumRegister(3, 'qreg2'), 2)), clbits=())"
+
+print(circuit.draw())
+>>> qreg1_0: ────────────────
+>>>          ┌───┐┌────────┐
+>>> qreg1_1: ┤ S ├┤0       ├─
+>>>          └───┘│        │
+>>> qreg1_2: ─────┤        ├─
+>>>               │        │
+>>> qreg2_0: ─────┤  Block ├─
+>>>          ┌───┐│        │
+>>> qreg2_1: ┤ H ├┤        ├─
+>>>          └───┘│        │
+>>> qreg2_2: ─────┤1       ├─
+>>>               └────────┘
+```
+
+## Outstanding questions:
+1. In the existing implementation of blocks that exists in `pec-runtime`, blocks are hashable, so that they can be used as keys in dictionaries (specifically in the dictionary porduced by the `NoiseLearner` that maps blocks to noise models). This is clearly unsafe since circuits are mutable. How can we design the noise handle so that it is safer than that, but also more flexible (e.g., if users want to mutate the block but still assign the same noise to it during mitigation, they should be allowed to do so)?
+2. Should `Block` be added to qiskit core, alongside the transpiler pass groups gates in blocks workflow 1? If so, should they be written in Rust or Python?
+3. What packages and subpackages need to be modified once `Block` is introduced? For example, should we add logic to qiskit-aer to be able to simulate circuits with blocks, or to Qiskit's optimizer, ..?
+4. How and when do we communicate the "scope of a block context"? For example, suppose a block context is only sensibly defined for depth-1 blocks of entanglers. Do we let this thing be instantiated with anything and raise during execution, or do we raise immediately? Who is in charge of validation?
+5. How can we (should we?) facilitate re-use of CircuitInstruction or Block instances? For example, suppose we want the same block in a circuit 100 times:
+```python
+for _ in range(100):
+    with circuit.block(context=my_context, name="foo") as block:
+        block.h(0) 
+        block.cx(0, 1)
+```
```

Comment:
4. Issue 4 can be more generally stated as: can the annotations of a box have an impact on the allowed contents of a box at construction time? Offline, a subset of us discussed whether or not this kind of thing would be more appropriate during primitive validation.

5. I'm not sure exactly what you mean by reuse. The Box concept being discussed is reusable in the sense that, like every other kind of instruction, there would be nothing stopping someone from adding the same box instance multiple times to a single circuit. By reuse, do you mean that tools such as the transpiler would be forced to perform the exact same operation on every like block encountered? Like if it scheduled the contents of a box in one way, if it ever encountered the same block again, it would by contract have to schedule the new one in the same way?